### ISSUE TITLE
gradients of tf.fake_quant_with_min_max_vars function.

### ISSUE THREAD

**Issue by sicnarf1a** [tensorflow/tensorflow#22977]
### System information
- **OS Platform and Distribution : Linux Ubuntu 16.04**
- **TensorFlow version : 1.6.0**
- **Python version : 3.6**
- **CUDA/cuDNN version : CUDA Version 9.0.176/ CUDNN 7.0.5**:
- **GPU model and memory : TITAN Xp/12196MiB**

### Describe the problem
I assume that **tf.fake_quant_with_min_max_vars function** can not be differentiable due to the fact that quantization should be working based on threshold, such as round, or sign function. And it means that we can not get the gradient of variables due to the nature of chain rule.

In these days, several tricks are researched, and one of the most popular method is called **'straight-through-estimator'**, literally passing through the gradient itself.

So I tested **tf.fake_quant_with_min_max_vars** to apply **'straight-through-estimator'** functionality. Most results get a gradient 1. This means it works well. Sometimes, however, the result is incorrect. More detailed errors are described below.

### Source code / logs
Here's some snippet code
`x = tf.cast(np.random.normal(0, 1, (10), tf.float32)`
`x_q = tf.fake_quant_with_min_max_vars(x, min=tf.reduce_min(x), max=tf.reduce_max(x), num_bits=3)`
`grad = tf.gradients(x_q, x)`

In that case, sometimes **grad** get weird results:
`[array([1., 1., 1., 0., 1., 1., 2., 1., 1., 1.], dtype=float32)]`
And non-zero gradient values have a common quantization value. In this case, the **x_q** is:
`array([0., 0., 0., -2.289673, 2.289673, 0., -2.289673, 0., 4.570346, 0.], dtype=float32)`

**Comment by sicnarf1a** [comment]
I forgot to mention that this error or bug is important because it has a lot of impact on convergence in the training process when using **tf.fake_quant_with_min_max_vars function**.

**Comment by michaelklachko** [comment]
I can reproduce it using tf.fake_quant_with_min_max_args function on Ubuntu 16.04 with python 2.7.12, numpy 1.14.0, and TF 1.11.0 (CPU only):

```
michael@Pascal:~$ cat tf_quant.py
import tensorflow as tf
import numpy as np

np.set_printoptions(precision=3, linewidth=120, suppress=True)

sess = tf.Session()

for i in range(3):
    sample = np.random.normal(0,1,(8,8))
    x = tf.cast(sample, tf.float32)

    #min_ = tf.reduce_min(x)
    #max_ = tf.reduce_max(x)
    min_ = np.min(sample)
    max_ = np.max(sample)
    x_q = tf.fake_quant_with_min_max_args(x, min=min_, max=max_, num_bits=2)

    grad = tf.gradients(x_q, x)
    print sess.run([x, x_q, grad])
```
```
michael@Pascal:~$ python2 tf_quant.py
2018-10-22 09:03:10.722455: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
       [array([
       [-0.418, -0.105,  0.994,  0.022, -1.241, -0.324,  0.128,  0.001],
       [ 0.399, -1.085,  0.361,  2.381,  0.095, -0.23 , -0.892, -0.864],
       [ 2.527,  0.456, -1.682,  0.049,  0.508, -0.515,  0.379, -0.514],
       [-1.474, -2.128,  1.136, -0.21 , -1.141, -0.059,  0.609,  1.454],
       [ 2.644, -0.136,  1.267,  0.397, -0.688, -1.37 , -0.625, -0.144],
       [-0.887, -1.912, -1.888,  0.066,  0.07 ,  0.166, -0.582, -1.693],
       [-0.032, -1.16 , -0.344, -0.31 ,  0.72 , -0.713,  0.756,  1.   ],
       [ 0.599, -2.627,  0.643,  2.024, -0.593,  0.528,  0.62 , -0.893]], dtype=float32), 
       array([
       [ 0.   ,  0.   ,  1.757,  0.   , -1.757,  0.   ,  0.   ,  0.   ],
       [ 0.   , -1.757,  0.   ,  1.757,  0.   ,  0.   , -1.757,  0.   ],
       [ 1.757,  0.   , -1.757,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [-1.757, -1.757,  1.757,  0.   , -1.757,  0.   ,  0.   ,  1.757],
       [ 3.514,  0.   ,  1.757,  0.   ,  0.   , -1.757,  0.   ,  0.   ],
       [-1.757, -1.757, -1.757,  0.   ,  0.   ,  0.   ,  0.   , -1.757],
       [ 0.   , -1.757,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  1.757],
       [ 0.   , -1.757,  0.   ,  1.757,  0.   ,  0.   ,  0.   , -1.757]], dtype=float32), 
       [array([
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 0., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 1., 1., 1., 1., 1., 1.]], dtype=float32)]]
       [array([
       [-0.367, -0.653,  0.494,  0.604,  3.324,  0.023,  2.17 ,  0.203],
       [-0.639,  0.214,  3.579, -0.498,  0.859,  2.193, -1.086, -0.746],
       [ 0.066,  1.084,  0.795,  1.373,  0.489,  0.945,  0.157,  0.013],
       [ 0.713,  0.306,  0.762,  0.298,  0.482, -1.181, -1.783, -0.756],
       [-0.658, -1.185,  0.671,  1.643,  0.282,  0.879, -0.78 ,  1.086],
       [ 0.031,  0.602, -0.686, -0.006, -1.228, -1.294, -1.03 ,  0.959],
       [-0.603,  0.039,  0.034, -0.62 ,  0.174,  0.121, -1.015,  0.202],
       [-0.544, -2.412, -1.444, -1.354, -0.576,  0.108,  1.627,  0.868]], dtype=float32), 
       array([
       [ 0.   ,  0.   ,  0.   ,  0.   ,  3.994,  0.   ,  1.997,  0.   ],
       [ 0.   ,  0.   ,  3.994,  0.   ,  0.   ,  1.997, -1.997,  0.   ],
       [ 0.   ,  1.997,  0.   ,  1.997,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   , -1.997, -1.997,  0.   ],
       [ 0.   , -1.997,  0.   ,  1.997,  0.   ,  0.   ,  0.   ,  1.997],
       [ 0.   ,  0.   ,  0.   ,  0.   , -1.997, -1.997, -1.997,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   , -1.997,  0.   ],
       [ 0.   , -1.997, -1.997, -1.997,  0.   ,  0.   ,  1.997,  0.   ]], dtype=float32), 
       [array([
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 1., 1., 1., 1., 1., 1.]], dtype=float32)]]
       [array([
       [-1.025,  0.599, -1.482,  0.203,  0.043,  0.098,  0.331, -2.39 ],
       [-0.097,  1.226, -0.622,  0.668,  0.386,  1.166, -1.397,  0.325],
       [ 1.481,  0.71 , -0.901,  0.784, -1.656,  0.436, -1.389, -1.009],
       [ 1.175,  0.308,  0.249, -0.82 , -2.894, -0.62 ,  0.989,  0.968],
       [ 0.775,  1.358,  0.506, -2.297, -0.183, -0.976, -0.743,  0.661],
       [ 0.09 ,  0.324, -1.645, -0.832,  1.265,  0.257,  0.999, -0.491],
       [ 2.886,  0.406,  0.393, -0.557, -0.68 , -0.243,  0.547, -1.332],
       [ 0.49 ,  0.412,  0.036, -0.434,  0.022,  0.679, -0.229,  1.001]], dtype=float32), 
       array([
       [-1.927,  0.   , -1.927,  0.   ,  0.   ,  0.   ,  0.   , -1.927],
       [ 0.   ,  1.927,  0.   ,  0.   ,  0.   ,  1.927, -1.927,  0.   ],
       [ 1.927,  0.   ,  0.   ,  0.   , -1.927,  0.   , -1.927, -1.927],
       [ 1.927,  0.   ,  0.   ,  0.   , -3.853,  0.   ,  1.927,  1.927],
       [ 0.   ,  1.927,  0.   , -1.927,  0.   , -1.927,  0.   ,  0.   ],
       [ 0.   ,  0.   , -1.927,  0.   ,  1.927,  0.   ,  1.927,  0.   ],
       [ 1.927,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   , -1.927],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  1.927]], dtype=float32), 
       [array([
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [0., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.]], dtype=float32)]]
```

**Comment by wt-huang** [comment]
@michaelklachko I run the code in your environment and results are normal. Note that you need to use `tf.reduce_max` and `tf.reduce_min` in `tf.fake_quant_with_min_max_vars()`

```
import tensorflow as tf
import numpy as np

for i in range(3):
    tf.reset_default_graph()
    x = tf.cast(np.random.normal(0, 1, (8,8)), tf.float32)
    x_q = tf.fake_quant_with_min_max_vars(x, min=tf.reduce_min(x), max=tf.reduce_max(x), num_bits=3)
    grad = tf.gradients(x_q, x)
    sess = tf.Session()
    print(sess.run([grad]))
```
```
[[array([[1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.]], dtype=float32)]]
[[array([[1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.]], dtype=float32)]]
[[array([[1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.]], dtype=float32)]]
```

**Comment by michaelklachko** [comment]
@wt-huang not sure I understand: have you ran my code exactly as shown? Did you not see any zero gradients?

**Comment by wt-huang** [comment]
@michaelklachko You are not calling the same function used in the code snippet from @sicnarf1a. I fixed your code so it would produce results as expected. Best to open a new issue so we can focus on your scenario.

**Comment by michaelklachko** [comment]
@wt-huang I think it's pretty obvious that the issue that caused the OP to open this ticket is the same one reproduced with my code. Now you have two people reporting the same issue, using different functions.

**Comment by wt-huang** [comment]
@michaelklachko `tf.fake_quant_with_min_max_vars` and `tf.fake_quant_with_min_max_args` are related but they carry different names so at least the title needs to be updated if we were to include them in one ticket. 

In your code snippet, need to use `tf.reduce_max` and `tf.reduce_min` for max and min respectively. Then you will get 1's all the way through. This is also what @sicnarf1a did.

Both `tf.fake_quant_with_min_max_vars` and `tf.fake_quant_with_min_max_args` can work if using my code snippet above. In the future we will certainly do our best to include more test cases and make the functions more user friendly.

**Comment by michaelklachko** [comment]
@wt-huang Now I'm really confused: are you saying the arguments min/max to `tf.fake_quant_with_min_max_args` should NOT be floats, as specified here: https://www.tensorflow.org/api_docs/python/tf/quantization/fake_quant_with_min_max_args ? Then what are the default values for that function?

**Comment by wt-huang** [comment]
The arguments min/max to `tf.fake_quant_with_min_max_args` should be floats.

**Comment by michaelklachko** [comment]
So are you ready to fix this bug, or should we keep splitting hairs?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @wt-huang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by zhly0** [comment]
@wt-huang ，Hi wt-huang：
    Is there a doc that can elaborate how to add fake node manually？Since in tensorflow 1.11,some op like transpose conv can not add fake node automatic.
Thanks,if you can help me,I search the internet for several days,but counld not find any doc.

**Comment by gadagashwini** [comment]
Hi @sicnarf1a, I could able execute the code successfully, output is pretty consistent.
I tried it on latest Tensorflow v2.9.

```
import tensorflow as tf
import numpy as np

for i in range(3):
    x = tf.cast(np.random.normal(0, 1, (8,8)), tf.float32)    
    with tf.GradientTape() as g:
      g.watch(x)
      x_q = tf.quantization.fake_quant_with_min_max_vars(x, min=tf.reduce_min(x), max=tf.reduce_max(x), num_bits=3)
      dy_dx = g.gradient(x_q, x)
    print(dy_dx)
```
```
tf.Tensor(
[[1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 0. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 3. 1. 1.]
 [1. 1. 1. 1. 0. 1. 1. 1.]], shape=(8, 8), dtype=float32)
tf.Tensor(
[[1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]], shape=(8, 8), dtype=float32)
tf.Tensor(
[[1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]], shape=(8, 8), dtype=float32)
```

**Comment by gadagashwini** [comment]
@sicnarf1a, Closing this issue, since it is resolved. Please feel free to reopen if you need more investigation. Thank you!
